src/dune_pkg/rev_store.ml

```diff
@@ -2,43 +2,101 @@ open Stdune
 open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
+module Scheduler = Dune_engine.Scheduler
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+type t =
+  { dir : Path.t
+  ; lock : Path.t
+  }
+
 type rev = Rev of string
 
-let equal { dir } t = Path.equal dir t.dir
+let lock_limited flock lock timeout =
+  let sleep_duration = 0.1 in
+  let res = ref (Ok `Failure) in
+  let+ () =
+    Fiber.repeat_while
+      ~f:(fun timeout ->
+        let timeout = max timeout 0. in
+        if timeout > 0.
+        then (
+          match Flock.lock_non_block flock lock with
+          | Ok `Success ->
+            res := Ok `Success;
+            Fiber.return None
+          | Ok `Failure ->
+            let+ () = Scheduler.sleep sleep_duration in
+            Some (timeout -. sleep_duration)
+          | err ->
+            res := err;
+            Fiber.return None)
+        else (
+          res := Ok `Failure;
+          Fiber.return None))
+      ~init:timeout
+  in
+  !res
+;;
+
+let with_flock lock_path ~f =
+  let open Fiber.O in
+  let parent = Path.parent_exn lock_path in
+  Path.mkdir_p parent;
+  let fd = Unix.openfile (Path.to_string lock_path) [ Unix.O_CREAT; O_RDONLY ] 0o644 in
+  let flock = Flock.create fd in
+  let max_duration = 5. in
+  Fiber.finalize
+    ~finally:(fun () ->
+      (* closing the fd releases the flock *)
```

Comment:
Should we remove the lock file as well? If it's present, it's a bit hard to say if any program is taking the lock at the moment.

Reply:
We can but existence and non-existence of the file doesn't indicate holding the lock as it isn't an atomic operation, for that one needs to query e.g. `lsof`. But we can remove it as a hint to the user that no lock is meant to be held.